### PR TITLE
:cow: Draw outline for color in index 0 for inventory icons.

### PR DESCRIPTION
**Pull request by FluffyQuack**
Default outline drawing:
devilutionx 2021-04-01 14-36-08-326

Fixed outline drawing:
devilutionx 2021-04-01 14-35-27-918

Colour in index 0 in the palette is typically used for shadows. When drawing outlines, the game skips this colour so it avoids making outlines around shadows, but since the colour is also sometimes used on the object itself as the colour black this makes the outline not properly follow the shape of many CEL frames.

Inventory icons never have any shadows, so for those, we could make sure to always draw outlines for #0 pixels.

### CODE REVIEW

**Comment by qndel** [comment]
what about making `skipColorIndexZero` an optional argument with default false value?

**Comment by FluffyQuack** [comment]
> what about making `skipColorIndexZero` an optional argument with default false value?

I have a counter proposal! I'll make it true by default. That way its default behaviour matches the CL2 outline draw function, and also, once your DrawInv rewrite PR is accepted, then it's 2 calls to CelBlitOutlineTo() which needs to set skipColorIndexZero to false.

**Comment by qndel** [comment]
yeah I meant true, whatever is more common <derp>

**Comment by qndel** [comment]
whatever requires less code 🤪

File: Source/engine.cpp [hunk 1]
```diff
@@ -480,22 +480,24 @@ void CelBlitOutlineTo(CelOutputBuffer out, BYTE col, int sx, int sy, BYTE *pCelB
 				if (dst < out.end() && dst > out.begin()) {
 					if (dst >= out.end() - out.pitch()) {
 						while (width) {
-							if (*src++) {
+							if (!skipColorIndexZero || *src > 0) {
```
Comment (yuripourre) [inline Source/engine.cpp:483]: You could change less lines using: `skipColorIndexZero && *src++`

Comment (AJenbo) [reply Source/engine.cpp:483]: I actually prefer the current form in this PR as it doesn't do multiple types of operation in a single line.

Comment (FluffyQuack) [reply Source/engine.cpp:483]: Someone correct me if I'm wrong, but wouldn't that lead to invalid behaviour? The src pointer needs to increase by one every pixel, but if the "if" statement is `skipColorIndexZero && *src++` then I think what's after "&&" is never ran whenever skipColorIndexZero is false.

Comment (galaxyhaxz) [reply Source/engine.cpp:483]: You could make the suggestion work if you flip it i.e. `*src++ && skipColorIndexZero` But I think the way you have it now makes it the most legible, as post-increment in this case could be confusing.

Comment (yuripourre) [reply Source/engine.cpp:483]: @FluffyQuack you are right, I missed that.

**Review by yuripourre — APPROVED** [review]

**Comment by julealgon** [comment]
"BUT MY VANILLA HIGHLIGHTS!!! I'M A PRO DIABLO PLAYER, I NEED THE ORIGINAL HIGHLIGHT BEHAVIOR"

File: Source/engine.cpp [hunk 2]
```diff
@@ -480,22 +480,24 @@ void CelBlitOutlineTo(CelOutputBuffer out, BYTE col, int sx, int sy, BYTE *pCelB
 				if (dst < out.end() && dst > out.begin()) {
 					if (dst >= out.end() - out.pitch()) {
 						while (width) {
-							if (*src++) {
+							if (!skipColorIndexZero || *src > 0) {
 								dst[-out.pitch()] = col;
 								dst[-1] = col;
 								dst[1] = col;
 							}
+							src++;
```
Comment (julealgon) [inline Source/engine.cpp:488]: Shouldn't we be incrementing this before the if, like the original code?

File: Source/inv.cpp [hunk 3]
```diff
@@ -185,9 +185,9 @@ void DrawInv(CelOutputBuffer out)
 				color = ICOL_RED;
 			}
 			if (frame <= 179) {
-				CelBlitOutlineTo(out, color, RIGHT_PANEL_X + 133, 59, pCursCels, frame, frame_width);
+				CelBlitOutlineTo(out, color, RIGHT_PANEL_X + 133, 59, pCursCels, frame, frame_width, false);
 			} else {
-				CelBlitOutlineTo(out, color, RIGHT_PANEL_X + 133, 59, pCursCels2, frame - 179, frame_width);
+				CelBlitOutlineTo(out, color, RIGHT_PANEL_X + 133, 59, pCursCels2, frame - 179, frame_width, false);
```
Comment (julealgon) [inline Source/inv.cpp:190]: Making it `false` by default would help indeed, considering most cases rely on it.

**Review by julealgon — APPROVED** [review]

File: Source/engine.cpp (hunk 2 above)
Comment (AJenbo) [reply Source/engine.cpp:488]: It's a post-increment, so it was not before the if but after... this is why I don't like having them inside other operations, everyone gets confused :joy:

Comment (julealgon) [reply Source/engine.cpp:488]: Oh... indeed. I'm dumb.

File: Source/inv.cpp (hunk 3 above)
Comment (FluffyQuack) [reply Source/inv.cpp:190]: https://github.com/diasurgical/devilutionX/pull/1306

Once that PR is accepted, then it's only 2 or 3 calls to the function which needs the argument to be set to false.

**Review by AJenbo — APPROVED** [review]

**Comment by AJenbo** [comment]
This one is too nice and clean to not include in 1.2.0 :+1:

**Pull request merged**